fix947_FreeBSD_BasicError_typeinfo

clanmills opened the pull request:
Fix #947

clanmills committed:
fix947_FreeBSD_BasicError_typeinfo

D4N commented on include/exiv2/error.hpp:
```diff
@@ -260,7 +260,7 @@ namespace Exiv2 {
      */
     template<typename charT>
-    class BasicError : public AnyError {
+    class EXIV2API BasicError : public AnyError {
```
This breaks builds on Windows and probably opens quite the can of worms: declaring a template as export/visible is not really supported.

piponazo replied:
Right ... the last time I worked on this topic we reached the same conclusion ... but somehow we always forget about it when somebody opens an issue in github related with that part of the code. 

I guess the final fix for this will be to finish the work started in #685

clanmills replied:
Oh gosh, this isn't good news.  The changes in #685 are huge.  Because they will change the API, I'd prefer not to do that in 0.27-maintenance.

I don't think we're happy with the following hack:
```
#ifdef __FREEBSD___
class EXIV2API BasicError : public AnyError
else
class BasicError : public AnyError
#endif
```
And we should do something similar on the change in src/CMakeLists.txt

For sure, we're not putting this into v0.27.2.  We need to explore other possibilities.  @lantw44 can you comment on this please?

lantw44 replied:
Does it work if we drop `EXIV2API` from `setMsg` in include/exiv2/error.hpp, and move "free functions, template and inline definitions" to src/error.cpp or mark them as `inline`?

clanmills replied:
Useful suggestions, thanks.  I'll wait to see if other in Team Exiv2 wish to comment and I'll investigate this evening.

I've created a fresh FreeBSD VM this morning to investigate your thoughts about not using the package manager to install GCC.

D4N replied:
> Does it work if we drop `EXIV2API` from `setMsg` in include/exiv2/error.hpp, and move "free functions, template and inline definitions" to src/error.cpp or mark them as `inline`?

`setMsg` must stay in the header because it is used in the constructor of `BasicError` and because of that it must be marked as `EXIV2API` as it otherwise will not be visible on Windows. Marking it inline is worth a try though, as that's probably the only option beside #685.

clanmills commented:
Bad News. 

This is inspired by Dan's "inline" idea.  There are 4 constructors for BasicError, and they all call setMsg().  I've totally removed setMsg() and placed the code in-line in the declaration of the constructors, and declared a free function `errMsg()` in error.hpp and defined it in error.cpp.

jens.cpp still aborts.

```
[rmills@rmillsmm-freebsd2 ~/temp]$ cat jens.cpp
// file: jens.cpp
// https://github.com/Exiv2/exiv2/issues/644#issuecomment-452472029

/*
Linux:
$ c++ -std=c++98 -I/usr/local/include -L/usr/local/lib -lexiv2 jens.cpp -o jens
$ ./jens
Caught error No namespace info available for XMP prefix `gnome'
http://www.gnome.org/xmp/
$
*/

#include <exiv2/exiv2.hpp>
#include <iostream>

int main(int argc, char *argv[])
{
        try {
                Exiv2::XmpProperties::ns("gnome");
        } catch (Exiv2::Error &error) {
                std::cerr << "Caught error " << error.what() << "\n";
                Exiv2::XmpProperties::registerNs("http://www.gnome.org/xmp", "gnome");
        }
        std::cout << Exiv2::XmpProperties::ns("gnome") << std::endl;
}
[rmills@rmillsmm-freebsd2 ~/temp]$ c++ -std=c++98 -I/usr/local/include -L/usr/local/lib -lexiv2 jens.cpp -o jens
[rmills@rmillsmm-freebsd2 ~/temp]$ ./jens
Abort trap (core dumped)
[rmills@rmillsmm-freebsd2 ~/temp]$ c++ -I/usr/local/include -L/usr/local/lib -lexiv2 jens.cpp -o jens
[rmills@rmillsmm-freebsd2 ~/temp]$ ./jens
Abort trap (core dumped)
[rmills@rmillsmm-freebsd2 ~/temp]$ 
```

I have run out of time to work on this for v0.27.2.  I'll review my efforts this evening.  If I don't have any new ideas, I will defer this for v0.27.3.  @lantw44, you'll have to use your work-around in the CMake code for now.

clanmills commented:
I've run ./jens through GDB:
```
root@rmillsmm-freebsd2:/home/rmills/temp # gdb jens
GNU gdb (GDB) 8.3 [GDB v8.3 for FreeBSD]
...
Reading symbols from jens...
(gdb) run
Starting program: /usr/home/rmills/temp/jens 

Program received signal SIGABRT, Aborted.
0x0000000800a4345a in thr_kill () from /lib/libc.so.7
(gdb) bt
#0  0x0000000800a4345a in thr_kill () from /lib/libc.so.7
#1  0x0000000800a41844 in raise () from /lib/libc.so.7
#2  0x00000008009b4079 in abort () from /lib/libc.so.7
#3  0x000000080083b8ba in ?? () from /lib/libcxxrt.so.1
#4  0x00000008004c5441 in Exiv2::XmpProperties::nsInfoUnsafe(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) () from /usr/local/lib/libexiv2.so.27
#5  0x00000008004c5296 in Exiv2::XmpProperties::ns(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&) () from /usr/local/lib/libexiv2.so.27
#6  0x0000000000202421 in main ()
(gdb)
```

```
    const XmpNsInfo* XmpProperties::nsInfoUnsafe(const std::string& prefix)
    {
        const XmpNsInfo::Prefix pf(prefix);
        const XmpNsInfo* xn = lookupNsRegistryUnsafe(pf);
        if (!xn) xn = find(xmpNsInfo, pf);
        if (!xn) std::cout << "\n\ntrouble in nsInfoUnsafe\n\n";
        if (!xn) throw Error(kerNoNamespaceInfoForXmpPrefix, prefix);
        return xn;
    }
```

It's not throwing the error.  The `Error(int,std::string)` constructor is causing libc to terminate the program.

```
[rmills@rmillsmm-freebsd2 ~/temp]$ c++ -I/usr/local/include -L/usr/local/lib -lexiv2 jens.cpp -o jens
[rmills@rmillsmm-freebsd2 ~/temp]$ ./jens 


trouble in nsInfoUnsafe

Abort trap (core dumped)
[rmills@rmillsmm-freebsd2 ~/temp]$
```

lantw44 commented:
Does this patch work for you?
```diff
diff --git a/include/exiv2/error.hpp b/include/exiv2/error.hpp
index 42a0959c..1787f10c 100644
--- a/include/exiv2/error.hpp
+++ b/include/exiv2/error.hpp
@@ -260,43 +260,43 @@ namespace Exiv2 {
              provided to print errors to a stream.
      */
     template<typename charT>
-    class BasicError : public AnyError {
+    class EXIV2API BasicError : public AnyError {
     public:
         //! @name Creators
         //@{
         //! Constructor taking only an error code
-        explicit BasicError(ErrorCode code);
+        explicit inline BasicError(ErrorCode code);
 
         //! Constructor taking an error code and one argument
         template<typename A>
-        BasicError(ErrorCode code, const A& arg1);
+        inline BasicError(ErrorCode code, const A& arg1);
 
         //! Constructor taking an error code and two arguments
         template<typename A, typename B>
-        BasicError(ErrorCode code, const A& arg1, const B& arg2);
+        inline BasicError(ErrorCode code, const A& arg1, const B& arg2);
 
         //! Constructor taking an error code and three arguments
         template<typename A, typename B, typename C>
-        BasicError(ErrorCode code, const A& arg1, const B& arg2, const C& arg3);
+        inline BasicError(ErrorCode code, const A& arg1, const B& arg2, const C& arg3);
 
         //! Virtual destructor. (Needed because of throw())
-        virtual ~BasicError() throw();
+        virtual inline ~BasicError() throw();
         //@}
 
         //! @name Accessors
         //@{
-        virtual int code() const throw();
+        virtual inline int code() const throw();
         /*!
           @brief Return the error message as a C-string. The pointer returned by what()
                  is valid only as long as the BasicError object exists.
          */
-        virtual const char* what() const throw();
+        virtual inline const char* what() const throw();
 #ifdef EXV_UNICODE_PATH
         /*!
           @brief Return the error message as a wchar_t-string. The pointer returned by
                  wwhat() is valid only as long as the BasicError object exists.
          */
-        virtual const wchar_t* wwhat() const throw();
+        virtual inline const wchar_t* wwhat() const throw();
 #endif
         //@}
 
@@ -304,7 +304,7 @@ namespace Exiv2 {
         //! @name Manipulators
         //@{
         //! Assemble the error message from the arguments
-        EXIV2API void setMsg();
+        void setMsg();
         //@}
 
         // DATA
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 34087004..e1c735dc 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -145,10 +145,9 @@ endif()
 
 set_target_properties( exiv2lib_int PROPERTIES
     POSITION_INDEPENDENT_CODE ON
-    COMPILE_DEFINITIONS exiv2lib_STATIC
+    COMPILE_DEFINITIONS exiv2lib_EXPORTS
 )
 
-
 target_include_directories(exiv2lib_int PRIVATE ${ZLIB_INCLUDE_DIR})
 target_include_directories(exiv2lib PRIVATE ${ZLIB_INCLUDE_DIR})
 
diff --git a/src/error.cpp b/src/error.cpp
index c8e2d9e0..a47db000 100644
--- a/src/error.cpp
+++ b/src/error.cpp
@@ -225,7 +225,7 @@ namespace Exiv2 {
     }
 
     template<>
-    void BasicError<char>::setMsg()
+    void EXIV2API BasicError<char>::setMsg()
     {
         std::string msg = _(errMsg(code_));
         std::string::size_type pos;
@@ -262,7 +262,7 @@ namespace Exiv2 {
 
 #ifdef EXV_UNICODE_PATH
     template<>
-    void BasicError<wchar_t>::setMsg()
+    void EXIV2API BasicError<wchar_t>::setMsg()
     {
         std::string s = _(errMsg(code_));
         std::wstring wmsg(s.begin(), s.end());
```

clanmills commented:
Thank You for sending this.  This is building exiv2 and jens on FreeBSD, MacOS-X and Linux.  I haven't tried yet on Windows.

I'm tired and will look at this tomorrow.  For sure, this looks promising.  Thank You very much.

clanmills committed:
Pushing very helpful suggestion by @lantw44

clanmills commented:
This has failed to build several editions of Visual Studio (and passed Visual Studio 2010).  This is typical:
```
exiv2.lib(exiv2.dll) : error LNK2005: "public: __cdecl Exiv2::BasicError<char>::BasicError<char>(enum Exiv2::ErrorCode)" (??0?$BasicError@D@Exiv2@@QEAA@W4ErrorCode@1@@Z) already defined in crwimage_int.obj [C:\Users\rmills\gnu\github\exiv2\buildserver\build\unitTests\unit_tests.vcxproj]
```
I'm going to release Exiv2 v0.27.2-RC3 today without this fix.

phako commented:
That failure is odd, that's exactly what `inline` is suppose to help with.

phako commented:
Am I reading the CMakeList.txt wrong or are you linking exiv2 basically twice in the unit tests:

https://github.com/Exiv2/exiv2/blob/beab2898caeaecef5a276955f73248c759b0bb0c/unitTests/CMakeLists.txt#L21

and 

https://github.com/Exiv2/exiv2/blob/beab2898caeaecef5a276955f73248c759b0bb0c/unitTests/CMakeLists.txt#L29

lantw44 commented:
> Am I reading the CMakeList.txt wrong or are you linking exiv2 basically twice in the unit tests:

Yes, I guess the reason to do it this way is that unit tests use `Exiv2::Internal`, which isn't exported. I am preparing a patch to remove the dependency on `libexiv2.dll` for `unit_tests.exe` to prevent Visual Studio from reporting it as a duplicate symbol. This means we will compile the source code of `libexiv2.dll` twice because we have to link exiv2lib staticlly into `libexiv2.dll`.

piponazo commented:
> Am I reading the CMakeList.txt wrong or are you linking exiv2 basically twice in the unit tests:

This was done with the objective of being able to test some internal functions and classes which are not exported. It could be the reason of the error message:

```
exiv2.lib(exiv2.dll) : error LNK2005: "public: __cdecl Exiv2::BasicError<char>::BasicError<char>(enum Exiv2::ErrorCode)" (??0?$BasicError@D@Exiv2@@QEAA@W4ErrorCode@1@@Z) already defined in crwimage_int.obj [C:\Users\rmills\gnu\github\exiv2\buildserver\build\unitTests\unit_tests.vcxproj]
```

phako commented:
> > Am I reading the CMakeList.txt wrong or are you linking exiv2 basically twice in the unit tests:
> 
> This was done with the objective of being able to test some internal functions and classes which are not exported. It could be the reason of the error message:
> 
> ```
> exiv2.lib(exiv2.dll) : error LNK2005: "public: __cdecl Exiv2::BasicError<char>::BasicError<char>(enum Exiv2::ErrorCode)" (??0?$BasicError@D@Exiv2@@QEAA@W4ErrorCode@1@@Z) already defined in crwimage_int.obj [C:\Users\rmills\gnu\github\exiv2\buildserver\build\unitTests\unit_tests.vcxproj]
> ```

Yes, sorry, that's what I tried to say. Forgot to add that to my commen :)

lantw44 commented:
Updated patch:
```diff
diff --git a/include/exiv2/error.hpp b/include/exiv2/error.hpp
index 42a0959c..1787f10c 100644
--- a/include/exiv2/error.hpp
+++ b/include/exiv2/error.hpp
@@ -260,43 +260,43 @@ namespace Exiv2 {
              provided to print errors to a stream.
      */
     template<typename charT>
-    class BasicError : public AnyError {
+    class EXIV2API BasicError : public AnyError {
     public:
         //! @name Creators
         //@{
         //! Constructor taking only an error code
-        explicit BasicError(ErrorCode code);
+        explicit inline BasicError(ErrorCode code);
 
         //! Constructor taking an error code and one argument
         template<typename A>
-        BasicError(ErrorCode code, const A& arg1);
+        inline BasicError(ErrorCode code, const A& arg1);
 
         //! Constructor taking an error code and two arguments
         template<typename A, typename B>
-        BasicError(ErrorCode code, const A& arg1, const B& arg2);
+        inline BasicError(ErrorCode code, const A& arg1, const B& arg2);
 
         //! Constructor taking an error code and three arguments
         template<typename A, typename B, typename C>
-        BasicError(ErrorCode code, const A& arg1, const B& arg2, const C& arg3);
+        inline BasicError(ErrorCode code, const A& arg1, const B& arg2, const C& arg3);
 
         //! Virtual destructor. (Needed because of throw())
-        virtual ~BasicError() throw();
+        virtual inline ~BasicError() throw();
         //@}
 
         //! @name Accessors
         //@{
-        virtual int code() const throw();
+        virtual inline int code() const throw();
         /*!
           @brief Return the error message as a C-string. The pointer returned by what()
                  is valid only as long as the BasicError object exists.
          */
-        virtual const char* what() const throw();
+        virtual inline const char* what() const throw();
 #ifdef EXV_UNICODE_PATH
         /*!
           @brief Return the error message as a wchar_t-string. The pointer returned by
                  wwhat() is valid only as long as the BasicError object exists.
          */
-        virtual const wchar_t* wwhat() const throw();
+        virtual inline const wchar_t* wwhat() const throw();
 #endif
         //@}
 
@@ -304,7 +304,7 @@ namespace Exiv2 {
         //! @name Manipulators
         //@{
         //! Assemble the error message from the arguments
-        EXIV2API void setMsg();
+        void setMsg();
         //@}
 
         // DATA
diff --git a/src/CMakeLists.txt b/src/CMakeLists.txt
index 34087004..e1c735dc 100644
--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -145,10 +145,9 @@ endif()
 
 set_target_properties( exiv2lib_int PROPERTIES
     POSITION_INDEPENDENT_CODE ON
-    COMPILE_DEFINITIONS exiv2lib_STATIC
+    COMPILE_DEFINITIONS exiv2lib_EXPORTS
 )
 
-
 target_include_directories(exiv2lib_int PRIVATE ${ZLIB_INCLUDE_DIR})
 target_include_directories(exiv2lib PRIVATE ${ZLIB_INCLUDE_DIR})
 
diff --git a/src/error.cpp b/src/error.cpp
index c8e2d9e0..a47db000 100644
--- a/src/error.cpp
+++ b/src/error.cpp
@@ -225,7 +225,7 @@ namespace Exiv2 {
     }
 
     template<>
-    void BasicError<char>::setMsg()
+    void EXIV2API BasicError<char>::setMsg()
     {
         std::string msg = _(errMsg(code_));
         std::string::size_type pos;
@@ -262,7 +262,7 @@ namespace Exiv2 {
 
 #ifdef EXV_UNICODE_PATH
     template<>
-    void BasicError<wchar_t>::setMsg()
+    void EXIV2API BasicError<wchar_t>::setMsg()
     {
         std::string s = _(errMsg(code_));
         std::wstring wmsg(s.begin(), s.end());
diff --git a/unitTests/CMakeLists.txt b/unitTests/CMakeLists.txt
index 6dbee160..215118c9 100644
--- a/unitTests/CMakeLists.txt
+++ b/unitTests/CMakeLists.txt
@@ -1,5 +1,22 @@
 find_package(GTest REQUIRED)
 
+get_target_property(exiv2lib_SOURCES exiv2lib SOURCES)
+get_target_property(exiv2lib_int_SOURCES exiv2lib_int SOURCES)
+get_target_property(exiv2lib_COMPILE_DEFINITIONS exiv2lib COMPILE_DEFINITIONS)
+get_target_property(exiv2lib_INCLUDE_DIRECTORIES exiv2lib INCLUDE_DIRECTORIES)
+get_target_property(exiv2lib_LINK_LIBRARIES exiv2lib LINK_LIBRARIES)
+
+set(unit_tests_exiv2lib_SOURCES)
+foreach(source IN LISTS exiv2lib_SOURCES exiv2lib_int_SOURCES)
+    if(source MATCHES "\.(c|cpp|h|hpp)$")
+        if(source MATCHES ".*/.*")
+            list(APPEND unit_tests_exiv2lib_SOURCES "${source}")
+        else()
+            list(APPEND unit_tests_exiv2lib_SOURCES "../src/${source}")
+        endif()
+    endif()
+endforeach()
+
 add_executable(unit_tests mainTestRunner.cpp
     gtestwrapper.h
     test_types.cpp
@@ -14,13 +31,24 @@ add_executable(unit_tests mainTestRunner.cpp
     test_helper_functions.cpp
     test_slice.cpp
     test_image_int.cpp
-    $<TARGET_OBJECTS:exiv2lib_int>
+    ${unit_tests_exiv2lib_SOURCES}
+)
+
+target_compile_definitions(unit_tests
+    PRIVATE
+        exiv2lib_STATIC
+        ${exiv2lib_COMPILE_DEFINITIONS}
+)
+
+target_include_directories(unit_tests
+    PRIVATE
+        ${exiv2lib_INCLUDE_DIRECTORIES}
 )
 
 #TODO Use GTest::GTest once we upgrade the minimum CMake version required
 target_link_libraries(unit_tests
     PRIVATE
-        exiv2lib
+        ${exiv2lib_LINK_LIBRARIES}
         ${GTEST_BOTH_LIBRARIES}
 )
 
@@ -50,4 +78,4 @@ endif()
 
 if (MSVC)
     set_target_properties(unit_tests PROPERTIES LINK_FLAGS "/ignore:4099")
-endif()
\ No newline at end of file
+endif()
```

clanmills committed:
Looking better all the time, Gentlemen.  I believe the only changes between the patches are in unitTests/CMakeLists.txt.

This build passes the test suite and jens.cpp on MacOS-X.  I'll submit and see what the CI has to say.  With the changes only relating to unitTests (which I didn't run on FreeBSD), we should not have disturbed Linux/FreeBSD/jens.cpp.

clanmills commented:
Jenkins/Visual Studio 2017 is happy and unit tests have been run (MacOS-X, Linux, Cygwin, MinGW and MSVC).

When the CI lights turn GREEN, could somebody please **Approve** the change and it will go into Exiv2 v0.27.2-RC3 and I'll publish that later today.  I don't plan any code changes to 0.27.2 before release on 2019-07-31 (Two weeks today).

Great Work Everybody.  I'm going back to cutting the hedge and other garden chores!

piponazo commented on unitTests/CMakeLists.txt:
```diff
@@ -14,13 +31,24 @@ add_executable(unit_tests mainTestRunner.cpp
     test_slice.cpp
     test_image_int.cpp
-    $<TARGET_OBJECTS:exiv2lib_int>
+    ${unit_tests_exiv2lib_SOURCES}
```
I would not change this. The idea of using the OBJECT library is to not recompile things. With a CMake OBJECT library we just take the .o files compiled while generating `exiv2lib` and we reuse them in the test target. With this change, we would be compiling all these files twice.

lantw44 replied:
The reason of doing so is that Windows seems to distinguish between normal symbols and DLL-imported symbols. If I just take the object files of `exiv2lib_int` and `exiv2lib` and put them here, Visual Studio complains that symbols it expects to be provided by a DLL ends up being linked statically. Therefore, I use `exiv2lib_STATIC` macro to disable `dllimport` and `dllexport` attributes and compile the same source files again in order to generate object files which can be linked statically.

piponazo replied:
oh I see. Thanks for the feedback and the efforts for dealing with this issue :+1:

clanmills replied:
@lantw44  I support @piponazo in thanking you for dealing with this issue.  You (and @kevinbackhouse) are the best contributors outside of Team Exiv2 in 2019.  Thank you very much.

When I write the release notes, I like to acknowledge contributors by name.  Here's the 

*Exiv2 v0.27.2.2 = Exiv2 v0.27.2 RC2*

_Second Release Candidate for quarterly “dot” release of Exiv2 v0.27.2_

A huge thank you to Kevin for outstanding contributions to this release.
Thank You to Dan and Luis for their solid efforts on Exiv2.
Thank You to Andreas S for representing Team Exiv2 at LGM 2019.
Contributors: Jens, Nehal, Phil, Toni

I'd like to give **YOU** a _**"huge thank you also"**_ in v0.27.2 if you'd like to tell me your name!

@phako: You'll see I have already acknowledged you.  Thank You also for your contribution.

clanmills commented:
Folks:  This is failing the build on the Mac on the CI machine, but passes on my laptop when I build with:
```
CMAKE_OPTIONS="-DCMAKE_BUILD_TYPE=Release -DEXIV2_ENABLE_VIDEO=ON -DEXIV2_ENABLE_WEBREADY=ON -DEXIV2_BUILD_UNIT_TESTS=ON -DEXIV2_ENABLE_NLS=OFF -DEXIV2_ENABLE_CURL=ON"
```

Here's the transcript from the CI build machine:  https://travis-ci.org/Exiv2/exiv2/jobs/560032789

```
Scanning dependencies of target unit_tests
[ 12%] Building CXX object unitTests/CMakeFiles/unit_tests.dir/mainTestRunner.cpp.o
In file included from <built-in>:374:
<command line>:2:37: error: whitespace recommended after macro name [-Werror]
#define exiv2lib_COMPILE_DEFINITIONS-NOTFOUND 1
```

This somehow related to the new code:
```
+get_target_property(exiv2lib_COMPILE_DEFINITIONS exiv2lib COMPILE_DEFINITIONS)
...
+target_compile_definitions(unit_tests
+    PRIVATE
+        exiv2lib_STATIC
+        ${exiv2lib_COMPILE_DEFINITIONS}
+)
```

What's "-NOTFOUND"?  I think the negative "-" is causing the problem!  So close!

piponazo commented:
> What's "-NOTFOUND"? I think the negative "-" is causing the problem! So close!

For some reason, it seems that we do not have compile definitions for that target on Mac OSX. We should be able to identify that in the CMake code and only add them if they are available. I will push a commit in this branch with the needed change.

piponazo committed:
Use compile definitions only if they are available

clanmills commented:
Oh boy, are we having fun.  Luis has gone to bed (and why not).  GitLab is failing on CentOS saying:
```
File /var/cache/yum/x86_64/7/epel/metalink.xml does not exist
Could not parse metalink https://mirrors.fedoraproject.org/metalink?repo=epel-7&arch=x86_64 error was 
No repomd file
Uploading artifacts...
WARNING: build/: no matching files                 
ERROR: No files to upload                          
ERROR: Job failed: exit code 1
```
Eh? @D4N  Any ideas?
https://gitlab.com/D4N/exiv2/-/jobs/253807133

lantw44 commented:
> GitLab is failing on CentOS saying:

I think it is just a temporary networking issue. Retrying the build may fix it.

clanmills commented:
@piponazo I believe we're finally done on this.  Can you merge into 0.27-maintenance and I can get 0.27.2-RC2 finished.  Phew!

@lantw44 Don't forget to tell me your name, so I can thank you in the release notes.  My email is robin@clanmills.com
